File: Algorithm/Python/352. Data Stream as Disjoint Intervals.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-09-22 15:12:24
# @Last modified by:   WuLC
# @Last Modified time: 2016-09-22 22:19:47
# @Email: liangchaowu5@gmail.com


# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e



# the final result has to be in order 


# method 1, modify the intervals directly and sort the result
class SummaryRanges(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.inter = []

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        idx, merge_2 = None, False
        for i in xrange(len(self.inter)):
            if self.inter[i].start <= val <=self.inter[i].end:
                return
            if self.inter[i].start - val == 1:
                if idx == None:
                    self.inter[i].start = val
                    idx = i
                else:
                    self.inter[idx].end = self.inter[i].end
                    idx = i
                    merge_2 = True
                    break
            elif self.inter[i].end - val == -1:
                if idx == None:
                    self.inter[i].end = val
                    idx = i
                else:
                    self.inter[idx].start = self.inter[i].start
                    idx = i
                    merge_2 = True
                    break
        if idx == None:
            self.inter.append(Interval(s=val, e=val))
        if merge_2:
            self.inter.remove(self.inter[i])
             
    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        return sorted(self.inter, key = lambda x:x.start)

File: Algorithm/Java/352. Data Stream as Disjoint Intervals.java
/**
* Author: WuLC
* Date:   2017-04-20 14:58:29
* Last modified by:   WuLC
* Last Modified time: 2017-04-20 15:00:35
* Email: liangchaowu5@gmail.com
*/

// use TreeMap in Java, O(logn) time complexity
// referer: https://discuss.leetcode.com/topic/46887/java-solution-using-treemap-real-o-logn-per-adding


/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
 
public class SummaryRanges 
{
    private TreeMap<Integer, Interval> tree;
 
    public SummaryRanges() 
    {
        tree = new TreeMap<Integer, Interval>();
    }
    
    
    public void addNum(int val) 
    {
        if (tree.containsKey(val)) return;
        Integer lower = tree.lowerKey(val);  // must use Integer, cause there may be null type
        Integer higher = tree.higherKey(val);
        if (lower != null && higher != null && tree.get(lower).end + 1 == val && tree.get(higher).start == val + 1)
        {
            tree.get(lower).end = tree.get(higher).end;
            tree.remove(higher);
        }
        else if(higher != null && tree.get(higher).start == val + 1)
        {
            // can not modify start of interval directly, cause key is related to it
            tree.put(val, new Interval(val, tree.get(higher).end));
            tree.remove(higher);
        }
        else if(lower != null && tree.get(lower).end + 1 >= val)
        {
            tree.get(lower).end = Math.max(val, tree.get(lower).end);
        }
        else
        {
            tree.put(val, new Interval(val, val));
        }
    }
    
    
    public List<Interval> getIntervals() 
    {
        return new ArrayList<Interval>(tree.values());
    }
}
